pkg/cmd/server/startconfig/master_args.go

```diff
@@ -0,0 +1,490 @@
+	if err != nil {
+		return nil, err
+	}
+	kubeAddr, err := commandArgs.KubeConnectionArgs.GetKubernetesAddress(masterAddr)
+	if err != nil {
+		return nil, err
+	}
+	masterPublicAddr, err := commandArgs.GetMasterPublicAddress()
+	if err != nil {
+		return nil, err
+	}
+	kubePublicAddr, err := commandArgs.GetKubernetesPublicAddress()
+	if err != nil {
+		return nil, err
+	}
+	assetPublicAddr, err := commandArgs.GetAssetPublicAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	kubeCA, err := commandArgs.KubeConnectionArgs.GetKubeCA(defaultRootCAFilename(certDir, "ca"))
+	if err != nil {
+		return nil, err
+	}
+	kubeCredentials, err := commandArgs.KubeConnectionArgs.GetKubeCredentials(commandArgs.KubeConnectionArgs.GetKubernetesCertInfo())
+	if err != nil {
+		return nil, err
+	}
+
+	corsAllowedOrigins := []string{}
+	corsAllowedOrigins = append(corsAllowedOrigins, commandArgs.CORSAllowedOrigins...)
+	// always include the all-in-one server's web console as an allowed CORS origin
+	// always include localhost as an allowed CORS origin
+	// always include master public address as an allowed CORS origin
+	for _, origin := range []string{assetPublicAddr.Host, masterPublicAddr.Host, "localhost", "127.0.0.1"} {
+		corsAllowedOrigins = append(corsAllowedOrigins, origin)
+	}
+
+	etcdAddress, err := commandArgs.GetEtcdAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	var etcdConfig *configapi.EtcdConfig
+	if !commandArgs.EtcdAddr.Provided {
+		etcdConfig, err = commandArgs.BuildSerializeableEtcdConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+	var kubernetesMasterConfig *configapi.KubernetesMasterConfig
+	if !commandArgs.KubeConnectionArgs.KubernetesAddr.Provided && len(commandArgs.KubeConnectionArgs.ClientConfigLoadingRules.CommandLinePath) == 0 {
+		kubernetesMasterConfig, err = commandArgs.BuildSerializeableKubeMasterConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	config := &configapi.OpenShiftMasterConfig{
+		MasterBindAddress:   commandArgs.BindAddr.URL.Host,
+		MasterAddress:       masterAddr.String(),
+		MasterPublicAddress: masterPublicAddr.String(),
+
+		AssetBindAddress:   commandArgs.GetAssetBindAddress(),
+		AssetPublicAddress: assetPublicAddr.String(),
+		LogoutURI:          cmdutil.Env("OPENSHIFT_LOGOUT_URI", ""),
+
+		EtcdURL:    etcdAddress.String(),
+		EtcdConfig: etcdConfig,
+
+		KubernetesAddress:       kubeAddr.String(),
+		KubernetesPublicAddress: kubePublicAddr.String(),
+		KubernetesCA:            *kubeCA,
+		KubernetesMasterConfig:  kubernetesMasterConfig,
+
+		CORSAllowedOrigins: corsAllowedOrigins,
+
+		// TODO I'm pretty sure this should be split into the RootCA which is to check for valid certs and the actual cert and key file.
+		// The crypto implementation has these glommed together and I'm not yet sure how to split them apart
+		// CertificateAuthority will be used to mint client certificates in connections to the API.
+		RootCA: commandArgs.GetRootCAFile(),
```

Comment:
The specific CA cert and private key are *only* used for minting. Root CA bundles (any number of certificates in sequence in a crt file) are needed for the following things:
1) to validate client certs for API auth
2) to validate client certs for OAuth requestheader validation (in PR https://github.com/openshift/origin/pull/1246)
3) to publish as accepted ClientCAs when starting the server (must be union of 1 and 2)